Add scrollend feature flag and initial idl definitions

nmoucht opened the pull request:
#### 3e582555ece593cef2edc0f01c45dc0798a3690a
<pre>
Add scrollend feature flag and initial idl definitions
<a href="https://bugs.webkit.org/show_bug.cgi?id=295561">https://bugs.webkit.org/show_bug.cgi?id=295561</a>
<a href="https://rdar.apple.com/155270861">rdar://155270861</a>

Reviewed by Tim Nguyen.

Add scrollend feature flag and initial idl definitions.

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-event-handlers-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLBodyElement.idl:
* Source/WebCore/html/HTMLFrameSetElement.idl:
* Source/WebCore/page/VisualViewport.idl:

Canonical link: <a href="https://commits.webkit.org/297134@main">https://commits.webkit.org/297134@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e5112a7a4fba4e8c2958d7b030c689027aa7f50

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110646 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30305 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20738 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116672 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/60913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112609 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38894 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84141 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/60913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113594 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24743 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99627 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64582 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24109 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17766 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60467 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103137 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94131 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17825 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119462 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109200 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37686 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27991 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93096 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38060 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95897 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92920 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37944 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15684 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33663 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17854 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37582 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43054 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133475 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37244 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36063 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40583 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38952 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -6493,6 +6493,20 @@ ScrollToTextFragmentMarkingEnabled:
+ScrollendEventEnabled:
+  type: bool
+  status: testable
+  category: animation
```
I would use dom as category

nmoucht replied:
Fixed.

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -6493,6 +6493,20 @@ ScrollToTextFragmentMarkingEnabled:
+  type: bool
+  status: testable
+  category: animation
+  humanReadableName: "Scrollend event"
```
nit: I wouldn't capitalize the event name

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
```diff
@@ -341,6 +341,7 @@ namespace WebCore {
     macro(OfflineAudioContext) \
     macro(OffscreenCanvas) \
     macro(OffscreenCanvasRenderingContext2D) \
+    macro(onscrollend) \
```
nit: can you move this further down with the other lowercase on* event attributes?

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/VisualViewport.idl:
nit: the spec link in this file should be updated to: https://drafts.csswg.org/cssom-view/#the-visualviewport-interface

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/html/HTMLBodyElement.idl:
```diff
@@ -35,6 +35,7 @@
     [WindowEventHandler] attribute EventHandler onload;
     [WindowEventHandler] attribute EventHandler onresize;
     [WindowEventHandler] attribute EventHandler onscroll;
+    [WindowEventHandler, EnabledBySetting=ScrollendEventEnabled] attribute EventHandler onscrollend;
```
The second table on https://html.spec.whatwg.org/multipage/webappapis.html#event-handlers-on-elements,-document-objects,-and-window-objects says it shouldn't be part of this list.

This IDL file should probably link to: https://html.spec.whatwg.org/multipage/webappapis.html#window-reflecting-body-element-event-handler-set

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/html/HTMLFrameSetElement.idl:
```diff
@@ -30,6 +30,7 @@
     [WindowEventHandler] attribute EventHandler onload;
     [WindowEventHandler] attribute EventHandler onresize;
     [WindowEventHandler] attribute EventHandler onscroll;
+    [WindowEventHandler, EnabledBySetting=ScrollendEventEnabled] attribute EventHandler onscrollend;
```
Same comment here as for HTMLBodyElement

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/dom/Document.cpp:
I would prefer not introducing this boilerplate (ditto for Document.h) until we're going to use it.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/html/HTMLBodyElement.idl:
```diff
@@ -17,7 +17,8 @@
  */
-
+ 
+ // https://html.spec.whatwg.org/multipage/webappapis.html#window-reflecting-body-element-event-handler-set
```
nit: There's a small extra space in front of the `//`

nt1m replied:
Maybe put this next to the event listeners, instead of the top of the file.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/html/HTMLFrameSetElement.idl:
```diff
@@ -18,6 +18,7 @@
  * Boston, MA 02110-1301, USA.
  */
 
+// https://html.spec.whatwg.org/multipage/webappapis.html#window-reflecting-body-element-event-handler-set
```
ditto for this, this should probably be next to the event listeners, not at the top of the file

nmoucht replied:
Fixed.

nmoucht committed:
Add scrollend feature flag and initial idl definitions
https://bugs.webkit.org/show_bug.cgi?id=295561
rdar://155270861

Reviewed by Tim Nguyen.

Add scrollend feature flag and initial idl definitions.

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-event-handlers-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLBodyElement.idl:
* Source/WebCore/html/HTMLFrameSetElement.idl:
* Source/WebCore/page/VisualViewport.idl:

Canonical link: https://commits.webkit.org/297134@main
